collections/_article/d3d12-shaders-with-a-twist.md

```diff
@@ -0,0 +1,112 @@
+
+Mesa's SPIR-V to NIR code is patched in the Godot D3D12 renderer to do precisely that. The values of the SCs are set to some **carefully chosen magic sentinel value plus the SC id**, so in a later step we can tell if a literal value corresponds to an SC, and which one.
+
+Some extra NIR operations are added depending on the type of the SC. For instance, for a `float`, there's some bit casting so the base value can be treated as a plain 32-bit integer but the shader eventually sees it as a floating point value and thus does the right thing with it once patched.
+
+### 2. Prevention of NIR level optimization
+
+Mesa's NIR machinery is powerful. It can apply a variety of optimization passes to the NIR representation of a shader. That's awesome, but an issue for us because if we don't inhibit it somehow, we end up again with a "fossilized" shader, only that this time it's obeying some weird sentinel values instead of sensible SC values. Whatever the case, that's bad for our purposes.
+
+The core issues lies in the fact that NIR has a _load constant_ opcode (`nir_intrinsic_load_constant`), which is used to "load" the sentinel value into some virtual register, and that the myriad of optimization passes NIR is run through can optimize it out (if it's possible to pre-compute some ALU operation or however else take a static decision about the fate of that value, for instance.)
+
+The solution was simply to invent a new opcode (`nir_intrinsic_load_constant_non_opt`), which works exactly like the original one, but, by being in the end another one, is unkown to the optimization passes, which have no option but leaving it in place.
+
+### 3. Obtaining patchable DXIL
+
+The last thing we ask Mesa to do for us is generating DXIL from the nicely patched NIR we have at the moment. But our "deliverable" consists also of a table of the bit offsets to patch in the DXIL for each SC id.
+
+I patched the code that emits constants to the DXIL stream so that it recognizes cases of SCs if they match the sentinel value mask. The rest of the bits are the SC id. On each occurrence, the corresponding callback is leveraged to provide the information to the Godot D3D12 driver. (It's a bit more complicated than that, since the offset in the bitcode has later to be adjusted to match where the `DXBC` chunk ends up in the whole DXIL blob.)
+
+### 4. Creating pipelines with patched DXIL
+
+The Direct3D rendering driver is now able to apply whatever specialization constant values are needed to the already compiled shader, which was our initial goal.
+
+Now I have to mention that LLVM bitcode is an akward beast and that it's not trivial to patch a miserable integer due to variable bit-rate encoding. (Speaking of that, the former approach I implemented had a limit on the number of bits that could be patched, which imposed big limitations in the usable range of integers and imposed an elephant-sized epsilon for floating point ones; those have been happily lifted in the current one.)
```

Comment:
```suggestion
Now I have to mention that LLVM bitcode is an akward beast and that it's not trivial to patch a miserable integer due to variable bit-rate encoding. (Speaking of that, the former approach I implemented had a limit on the number of bits that could be patched. This imposed big limitations in the usable range of integers and imposed an elephant-sized epsilon for floating point ones; those have been happily lifted in the current one.)
```

Reply:
You helping me avoiding overly long sentences once more. 😀